BUG: Group-by numeric type-coercion with datetime

gwpdt opened the pull request:
closes #14423
closes #15421
closes #15670

During a group-by/apply on a DataFrame, in the presence of one or more
DateTime-like columns, Pandas would incorrectly coerce the type of all
other columns to numeric.  E.g. a String column would be coerced to
numeric, producing NaNs.

gwpdt committed:
BUG: Group-by numeric type-coericion with datetime

GH Bug #14423

During a group-by/apply on a DataFrame, in the presence of one or more
DateTime-like columns, Pandas would incorrectly coerce the type of all
other columns to numeric.  E.g. a String column would be coerced to
numeric, producing NaNs.

Fix the issue, and add a test.

gwpdt committed:
CLN: PEP8 (whitespace fixes)

jreback commented on pandas/core/groupby.py:
```diff
@@ -3566,7 +3568,8 @@ def first_non_None_value(values):
                 if (so.ndim == 2 and so.dtypes.apply(is_datetimelike).any()):
-                    result = result._convert(numeric=True)
+                    result = result.apply(
+                        lambda x: pd.to_numeric(x, errors='ignore'))
```
don't import pd
import to_numeric from pandas.util.misc (inside the function)

jreback replied:
or u can import pd inside the function is ok too

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -4314,6 +4314,16 @@ def test_cummin_cummax(self):
         tm.assert_series_equal(result, expected)
 
+    def test_numeric_coercion(self):
+        # GH 14423
```
there are a couple of issues marked duplicate can u add there tests here as well (inside this test is ok)

jreback replied:
add a 1-liner explaining this (maybe also make the test name a bit more descriptive)

jreback reviewed (changes requested):
lgtm couple of comments
pls add a whatsnew note in bug fixes
ping when green or need review

jreback commented:
pls add tests from these two

xref #14873 (boolean casts)
xref #14849 (datetime)

if this change doesn't fix these, pls lmk.

gwpdt commented:
I tried adding the test for #14873 - my fix does not resolve this issue.  It's similar, but unrelated.   It seems to be this line in _python_agg_general:

`output[name] = self._try_cast(result, obj, numeric_only=True)`

that seems to favour converting the result of the apply back to the same type as the original series.

jreback commented:
@gwpdt ok, we can treat that issue independently.

gwpdt commented:
I also tried adding the test for #14849 - my fix does not resolve this issue either.

The issue with that test is that it is testing for preservation of dtype="object" for a column whose values are all numeric (and therefore the "soft coercion" to numeric type succeeds).

In this instance (in the absence of any non-numeric values in the series), even with my fix, it will still convert the dtype of the column from "object" to numeric.

I tried removing the conversion altogether.  That fixes #14849, but breaks another test (test_time_field_bug / #11324), so I will not commit that change.

I still think it is worth contributing my fix, since it improves upon the current situation.  A series with string values will not be converted to NaNs when doing groupby/apply on a DF with one or more datetime columns.

I will address other review feedback (re: "pd" import) and update the PR.

gwpdt committed:
CLN: move import, add whatsnew entry

jreback commented:
AFAICT this *only* resolved #15421 (w/o breaking anything else). so thats good.

jreback commented:
#14423 I think *should* be fixable as well. want to give that a shot? (the example is not very good though).

gwpdt committed:
TST: Rename and expand test_numeric_coercion

Rename test_numeric_coercion to
test_apply_numeric_coercion_when_datetime, and add tests for GH #15421
and #14423

gwpdt commented:
Agreed - #15670 #15421 and #14423 are all fixed.

jreback commented:
lgtm. merging shortly.

jreback commented:
thanks!
